Add the time-weakening function

Huihuiweng opened the pull request:
The way to call time weakening in the input file "DATA/Par_file_faults" is to use: &RUPTURE_SWITCHES RATE_AND_STATE=.false., TWF=.true./. 
The way to set up the time weakening parameters is to add another line in "DATA/Par_file_faults": &TWF nuc_x=0, nuc_y=0, nuc_z=0, nuc_r=20000.0, nuc_t0=0.5, nuc_v=333.0/, where nuc_x, nuc_y, and nuc_z are the center of the time weakening, nuc_r is the nucleation radius, nuc_t0 is the cohesive time, and nuc_v is the nucleation speed.

Revision details:
In the file "fault_solver_common.f90", I defined the twf_type by:
  type twf_type
    real(kind=CUSTOM_REAL) ::  nuc_x, nuc_y, nuc_z, nuc_r, nuc_t0, nuc_v
  end type twf_type

If TWF is set as .true., then allocate bc%twf in the file "fault_solver_dynamic.f90" by:
  if (TWF)  then
    allocate(bc%twf)
    call twf_init(bc%twf,IIN_PAR)
  endif

  subroutine twf_init(f,IIN_PAR)
  implicit none
  type(twf_type), intent(out) :: f
  integer, intent(in) :: IIN_PAR
  integer :: ier
  real(kind=CUSTOM_REAL) :: nuc_x, nuc_y, nuc_z, nuc_r, nuc_t0, nuc_v
  NAMELIST / TWF / nuc_x, nuc_y, nuc_z, nuc_r,nuc_t0,nuc_v
  nuc_x  = 0.0e0_CUSTOM_REAL
  nuc_y  = 0.0e0_CUSTOM_REAL
  nuc_z  = 0.0e0_CUSTOM_REAL
  nuc_r  = 0.0e0_CUSTOM_REAL
  nuc_t0 = 0.0e0_CUSTOM_REAL
  nuc_v  = 0.0e0_CUSTOM_REAL
  read(IIN_PAR, nml=TWF,iostat=ier)
  if (ier /= 0) write(*,*) 'TWF not found in Par_file_faults.'
  f%nuc_x  = nuc_x
  f%nuc_y  = nuc_y
  f%nuc_z  = nuc_z
  f%nuc_r  = nuc_r
  f%nuc_t0 = nuc_t0
  f%nuc_v  = nuc_v
  end subroutine twf_init

Finally, the time weakening works by these codes:
      if (TWF) then
        timeval = it*bc%dt
        nuc_x   = bc%twf%nuc_x
        nuc_y   = bc%twf%nuc_y
        nuc_z   = bc%twf%nuc_z
        nuc_r   = bc%twf%nuc_r
        nuc_t0  = bc%twf%nuc_t0
        nuc_v   = bc%twf%nuc_v
        do i=1,bc%nglob
            dist = ((bc%coord(1,i)-nuc_x)**2 + (bc%coord(2,i)-nuc_y)**2 + (bc%coord(3,i)-nuc_z)**2)**0.5
            if (dist <= nuc_r) then
                tw_r     = timeval * nuc_v
                coh_size = nuc_t0  * nuc_v
                if(dist<= tw_r - coh_size) then
                    bc%MU(i) = min(bc%MU(i), bc%swf%mud(i))
                else if (dist> tw_r - coh_size .and. dist<= tw_r ) then
                    bc%MU(i) = min(bc%MU(i), bc%swf%mud(i) + (dist-(tw_r-coh_size))/coh_size*(bc%swf%mus(i)-bc%swf%mud(i)))
                endif
            endif
        enddo
      endif

komatits commented:
Thanks! Very useful.

Could you cut and paste that in the users manual?

Best wishes,
Dimitri.

On 11/6/18 5:37 PM, Huihui Weng wrote:
> The way to call time weakening in the input file "DATA/Par_file_faults" 
> is to use: &RUPTURE_SWITCHES RATE_AND_STATE=.false., TWF=.true./.
> The way to set up the time weakening parameters is to add another line 
> in "DATA/Par_file_faults": &TWF nuc_x=0, nuc_y=0, nuc_z=0, 
> nuc_r=20000.0, nuc_t0=0.5, nuc_v=333.0/, where nuc_x, nuc_y, and nuc_z 
> are the center of the time weakening, nuc_r is the nucleation radius, 
> nuc_t0 is the cohesive time, and nuc_v is the nucleation speed.
> 
> Revision details:
> In the file "fault_solver_common.f90", I defined the twf_type by:
> type twf_type
> real(kind=CUSTOM_REAL) :: nuc_x, nuc_y, nuc_z, nuc_r, nuc_t0, nuc_v
> end type twf_type
> 
> If TWF is set as .true., then allocate bc%twf in the file 
> "fault_solver_dynamic.f90" by:
> if (TWF) then
> allocate(bc%twf)
> call twf_init(bc%twf,IIN_PAR)
> endif
> 
> subroutine twf_init(f,IIN_PAR)
> implicit none
> type(twf_type), intent(out) :: f
> integer, intent(in) :: IIN_PAR
> integer :: ier
> real(kind=CUSTOM_REAL) :: nuc_x, nuc_y, nuc_z, nuc_r, nuc_t0, nuc_v
> NAMELIST / TWF / nuc_x, nuc_y, nuc_z, nuc_r,nuc_t0,nuc_v
> nuc_x = 0.0e0_CUSTOM_REAL
> nuc_y = 0.0e0_CUSTOM_REAL
> nuc_z = 0.0e0_CUSTOM_REAL
> nuc_r = 0.0e0_CUSTOM_REAL
> nuc_t0 = 0.0e0_CUSTOM_REAL
> nuc_v = 0.0e0_CUSTOM_REAL
> read(IIN_PAR, nml=TWF,iostat=ier)
> if (ier /= 0) write(/,/) 'TWF not found in Par_file_faults.'
> f%nuc_x = nuc_x
> f%nuc_y = nuc_y
> f%nuc_z = nuc_z
> f%nuc_r = nuc_r
> f%nuc_t0 = nuc_t0
> f%nuc_v = nuc_v
> end subroutine twf_init
> 
> Finally, the time weakening works by these codes:
> if (TWF) then
> timeval = it*bc%dt
> nuc_x = bc%twf%nuc_x
> nuc_y = bc%twf%nuc_y
> nuc_z = bc%twf%nuc_z
> nuc_r = bc%twf%nuc_r
> nuc_t0 = bc%twf%nuc_t0
> nuc_v = bc%twf%nuc_v
> do i=1,bc%nglob
> dist = ((bc%coord(1,i)-nuc_x)**2 + (bc%coord(2,i)-nuc_y)**2 + 
> (bc%coord(3,i)-nuc_z)**2)*/0.5
> if (dist <= nuc_r) then
> tw_r = timeval * nuc_v
> coh_size = nuc_t0 * nuc_v
> if(dist<= tw_r - coh_size) then
> bc%MU(i) = min(bc%MU(i), bc%swf%mud(i))
> else if (dist> tw_r - coh_size .and. dist<= tw_r ) then
> bc%MU(i) = min(bc%MU(i), bc%swf%mud(i) + 
> (dist-(tw_r-coh_size))/coh_size/(bc%swf%mus(i)-bc%swf%mud(i)))
> endif
> endif
> enddo
> endif
> 
> ------------------------------------------------------------------------
> 
> 
>         You can view, comment on, or merge this pull request online at:
> 
> https://github.com/geodynamics/specfem3d/pull/1312
> 
> 
>         Commit Summary
> 
>   * Add the time-weakening function
> 
> 
>         File Changes
> 
>   * *M* src/specfem3D/fault_solver_common.f90
>     <https://github.com/geodynamics/specfem3d/pull/1312/files#diff-0> (6)
>   * *M* src/specfem3D/fault_solver_dynamic.f90
>     <https://github.com/geodynamics/specfem3d/pull/1312/files#diff-1> (67)
> 
> 
>         Patch Links:
> 
>   * https://github.com/geodynamics/specfem3d/pull/1312.patch
>   * https://github.com/geodynamics/specfem3d/pull/1312.diff
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/geodynamics/specfem3d/pull/1312>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFjDKc-w7nTICCGsuxNcT5vmy4ZyRCsyks5usbqzgaJpZM4YQ0ru>.
> 

-- 
Dimitri Komatitsch, CNRS Research Director (DR CNRS)
Laboratory of Mechanics and Acoustics, Marseille, France
http://komatitsch.free.fr

Huihuiweng commented:
> Thanks! Very useful. Could you cut and paste that in the users manual? Best wishes, Dimitri.
> […](#)

I have added them in the users manual, and I also added a CUBIT meshing example. In this meshing example, there are a lot of descriptions for the usage. But in this meshing example, I cut the curved fault through the side boundaries, which seems to create an MPI error.  I will update this example later. Thanks.
